File: 2025/Day12/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Shape [][]int // list of orientations, each orientation is list of cell indices

// Parse shape from lines into cell coordinates
func parseShape(lines []string) [][2]int {
	var cells [][2]int
	for r, line := range lines {
		for c, ch := range line {
			if ch == '#' {
				cells = append(cells, [2]int{r, c})
			}
		}
	}
	return normalize(cells)
}

// Normalize so min row/col is 0
func normalize(cells [][2]int) [][2]int {
	if len(cells) == 0 {
		return cells
	}
	minR, minC := cells[0][0], cells[0][1]
	for _, c := range cells {
		if c[0] < minR {
			minR = c[0]
		}
		if c[1] < minC {
			minC = c[1]
		}
	}
	result := make([][2]int, len(cells))
	for i, c := range cells {
		result[i] = [2]int{c[0] - minR, c[1] - minC}
	}
	return result
}

// Generate all 8 orientations (4 rotations × 2 flips)
func allOrientations(cells [][2]int) [][][2]int {
	seen := make(map[string]bool)
	var result [][][2]int

	current := cells
	for i := 0; i < 4; i++ {
		// Original
		norm := normalize(current)
		key := fmt.Sprintf("%v", norm)
		if !seen[key] {
			seen[key] = true
			result = append(result, norm)
		}

		// Flipped
		flipped := make([][2]int, len(current))
		for j, c := range current {
			flipped[j] = [2]int{c[0], -c[1]}
		}
		norm = normalize(flipped)
		key = fmt.Sprintf("%v", norm)
		if !seen[key] {
			seen[key] = true
			result = append(result, norm)
		}

		// Rotate 90°
		rotated := make([][2]int, len(current))
		for j, c := range current {
			rotated[j] = [2]int{c[1], -c[0]}
		}
		current = rotated
	}
	return result
}

// Precompute valid placements as bitmasks
func validPlacements(orientations [][][2]int, h, w int) []uint64 {
	seen := make(map[uint64]bool)
	var placements []uint64

	for _, cells := range orientations {
		maxR, maxC := 0, 0
		for _, c := range cells {
			if c[0] > maxR {
				maxR = c[0]
			}
			if c[1] > maxC {
				maxC = c[1]
			}
		}

		for r := 0; r <= h-1-maxR; r++ {
			for c := 0; c <= w-1-maxC; c++ {
				var mask uint64
				for _, cell := range cells {
					bit := (r+cell[0])*w + (c + cell[1])
					mask |= 1 << bit
				}
				if !seen[mask] {
					seen[mask] = true
					placements = append(placements, mask)
				}
			}
		}
	}
	return placements
}

// Backtracking solver with bitmask
func solve(used uint64, placements [][]uint64, idx int) bool {
	if idx >= len(placements) {
		return true
	}

	for _, mask := range placements[idx] {
		if used&mask == 0 {
			if solve(used|mask, placements, idx+1) {
				return true
			}
		}
	}
	return false
}

// Check if shapes fit in region
func canFit(w, h int, shapes [][][][2]int, quantities []int) bool {
	var toPlace [][][][2]int
	totalCells := 0

	for idx, qty := range quantities {
		if idx >= len(shapes) || len(shapes[idx]) == 0 {
			continue
		}
		cellCount := len(shapes[idx][0])
		totalCells += qty * cellCount
		for i := 0; i < qty; i++ {
			toPlace = append(toPlace, shapes[idx])
		}
	}

	if len(toPlace) == 0 {
		return true
	}

	gridSize := w * h
	if totalCells > gridSize {
		return false
	}

	// For grids > 64 cells, fall back to area check only
	if gridSize > 64 {
		return true
	}

	// Compute placements for each shape
	var placementsList [][]uint64
	for _, shapeOrientations := range toPlace {
		p := validPlacements(shapeOrientations, h, w)
		if len(p) == 0 {
			return false
		}
		placementsList = append(placementsList, p)
	}

	// Sort by fewest placements first
	for i := 0; i < len(placementsList)-1; i++ {
		for j := i + 1; j < len(placementsList); j++ {
			if len(placementsList[j]) < len(placementsList[i]) {
				placementsList[i], placementsList[j] = placementsList[j], placementsList[i]
			}
		}
	}

	return solve(0, placementsList, 0)
}

func main() {
	filename := "input.txt"
	if len(os.Args) > 1 {
		filename = os.Args[1]
	}

	file, err := os.Open(filename)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}
	defer file.Close()

	var shapes [][][][2]int
	var regions [][3]interface{} // [w, h, quantities]

	scanner := bufio.NewScanner(file)
	var currentLines []string
	var currentIdx int = -1

	for scanner.Scan() {
		line := scanner.Text()

		if line == "" {
			if currentIdx >= 0 && len(currentLines) > 0 {
				// Extend shapes slice if needed
				for len(shapes) <= currentIdx {
					shapes = append(shapes, nil)
				}
				shapes[currentIdx] = allOrientations(parseShape(currentLines))
			}
			currentIdx = -1
			currentLines = nil
			continue
		}

		// Shape definition?
		if len(line) > 0 && line[0] >= '0' && line[0] <= '9' && strings.Contains(line, ":") {
			parts := strings.SplitN(line, ":", 2)
			idx, _ := strconv.Atoi(strings.TrimSpace(parts[0]))

			// Check if it's a region (WxH:)
			if strings.Contains(parts[0], "x") {
				dims := strings.Split(parts[0], "x")
				w, _ := strconv.Atoi(dims[0])
				h, _ := strconv.Atoi(dims[1])
				qParts := strings.Fields(strings.TrimSpace(parts[1]))
				var quantities []int
				for _, q := range qParts {
					n, _ := strconv.Atoi(q)
					quantities = append(quantities, n)
				}
				regions = append(regions, [3]interface{}{w, h, quantities})
			} else {
				// Shape definition
				currentIdx = idx
				rest := strings.TrimSpace(parts[1])
				if rest != "" {
					currentLines = []string{rest}
				} else {
					currentLines = nil
				}
			}
		} else if currentIdx >= 0 {
			currentLines = append(currentLines, line)
		} else if strings.Contains(line, "x") && strings.Contains(line, ":") {
			// Region on its own line
			parts := strings.SplitN(line, ":", 2)
			dims := strings.Split(strings.TrimSpace(parts[0]), "x")
			w, _ := strconv.Atoi(dims[0])
			h, _ := strconv.Atoi(dims[1])
			qParts := strings.Fields(strings.TrimSpace(parts[1]))
			var quantities []int
			for _, q := range qParts {
				n, _ := strconv.Atoi(q)
				quantities = append(quantities, n)
			}
			regions = append(regions, [3]interface{}{w, h, quantities})
		}
	}

	// Don't forget last shape
	if currentIdx >= 0 && len(currentLines) > 0 {
		for len(shapes) <= currentIdx {
			shapes = append(shapes, nil)
		}
		shapes[currentIdx] = allOrientations(parseShape(currentLines))
	}

	count := 0
	for i, region := range regions {
		w := region[0].(int)
		h := region[1].(int)
		q := region[2].([]int)

		if len(regions) > 10 {
			fmt.Fprintf(os.Stderr, "\r%d/%d", i+1, len(regions))
		}

		if canFit(w, h, shapes, q) {
			count++
		}
	}

	if len(regions) > 10 {
		fmt.Fprintln(os.Stderr)
	}

	fmt.Printf("Part 1: %d\n", count)
}

File: 2025/Day12/solution.rb
#!/usr/bin/env ruby
# Advent of Code 2025 - Day 12: Christmas Tree Farm
require 'set'
require 'timeout'

# Parse a shape from array of row strings
def parse_shape_lines(lines)
  cells = []
  lines.each_with_index do |row, r|
    row.strip.chars.each_with_index { |ch, c| cells << [r, c] if ch == '#' }
  end
  normalize(cells)
end

def normalize(cells)
  return [] if cells.empty?
  min_r = cells.map(&:first).min
  min_c = cells.map(&:last).min
  cells.map { |r, c| [r - min_r, c - min_c] }.sort
end

def all_orientations(cells)
  return [[]] if cells.empty?
  orientations = Set.new
  current = cells.dup
  4.times do
    orientations << normalize(current)
    orientations << normalize(current.map { |r, c| [r, -c] })
    current = current.map { |r, c| [c, -r] }
  end
  orientations.to_a
end

# Precompute all valid placements as bitmasks
def valid_placements(orientations, h, w)
  placements = []
  orientations.each do |cells|
    max_r = cells.map(&:first).max
    max_c = cells.map(&:last).max
    (0..(h - 1 - max_r)).each do |r|
      (0..(w - 1 - max_c)).each do |c|
        mask = cells.map { |dr, dc| (r + dr) * w + (c + dc) }
        placements << mask
      end
    end
  end
  placements.uniq
end

$timeout_flag = false

def solve_bitmask(used, placements_list, idx)
  return false if $timeout_flag
  return true if idx >= placements_list.size

  placements_list[idx].each do |mask|
    next if mask.any? { |bit| used[bit] }
    mask.each { |bit| used[bit] = true }
    return true if solve_bitmask(used, placements_list, idx + 1)
    mask.each { |bit| used[bit] = false }
  end
  false
end

def can_fit?(width, height, shapes, quantities)
  shapes_to_place = []
  quantities.each_with_index do |qty, idx|
    next if shapes[idx].nil? || shapes[idx].empty? || shapes[idx] == [[]]
    qty.times { shapes_to_place << shapes[idx] }
  end

  return true if shapes_to_place.empty?

  total_cells = shapes_to_place.sum { |orients| orients.first.size }
  grid_size = width * height
  return false if total_cells > grid_size

  placements_list = shapes_to_place.map { |orients| valid_placements(orients, height, width) }
  return false if placements_list.any?(&:empty?)
  placements_list.sort_by!(&:size)

  $timeout_flag = false
  used = Array.new(grid_size, false)
  
  begin
    Timeout.timeout(0.1) { solve_bitmask(used, placements_list, 0) }
  rescue Timeout::Error
    $timeout_flag = true
    false
  end
end

def solve_puzzle(filename)
  input = File.read(filename)
  blocks = input.strip.split(/\n\n+/)

  shapes = {}
  regions = []

  blocks.each do |block|
    lines = block.strip.split("\n")
    first = lines.first.strip

    if first =~ /^(\d+):(.*)$/
      idx = Regexp.last_match(1).to_i
      rest = Regexp.last_match(2).strip
      shape_lines = rest.empty? ? lines[1..] : [rest] + lines[1..]
      shapes[idx] = all_orientations(parse_shape_lines(shape_lines))
    else
      lines.each do |line|
        if line =~ /^(\d+)x(\d+):\s*(.*)$/
          w = Regexp.last_match(1).to_i
          h = Regexp.last_match(2).to_i
          q = Regexp.last_match(3).split.map(&:to_i)
          regions << [w, h, q]
        end
      end
    end
  end

  max_idx = shapes.keys.max || -1
  shapes_array = (0..max_idx).map { |i| shapes[i] || [[]] }

  count = 0
  regions.each_with_index do |(w, h, q), i|
    STDERR.print "\r#{i+1}/#{regions.size}" if regions.size > 10
    count += 1 if can_fit?(w, h, shapes_array, q)
  end
  STDERR.puts if regions.size > 10
  count
end

# Main
filename = ARGV[0] || "input.txt"
result = solve_puzzle(filename)
puts "Part 1: #{result}"
